doc/man/man3/sockpath.3.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man3/sockpath.3.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.\"
.\"
.TH SOCKPATH 3 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Library Calls"
.\"
.SH NAME
.B sockpath, setsockpath, getsockpath, endsockpath, freesockpathent, getsockpathent, sp_perror, sp_sperror
\- Socket Path database subroutines.
.\"
.\"
.SH SYNOPSIS
.HP
.B #include <sockdb.h>
.HP
.BI "void *setsockpath(void);"
.HP
.BI "struct sockpath *getsockpath(void *" handle );
.HP
.BI "struct sockpath *getsockpathent(int " domain ", int " type ", int " protocol );
.HP
.BI "void freesockpathent(struct sockpath *" sockpath );
.HP
.BI "int endsockpath(void *" handle );
.HP
.BI "void sp_perror(const char *" msg );
.HP
.BI "char *sp_sperror(void);"
.\"
.\"
.SH ARGUMENTS
.TP \w'\fIsockpath\fR\(em'u
.I handle
an opaque pointer returned by
.BR setsockpath ()
to be used as an argument to calls to
.BR getsockpath ().
.TP
.I domain
The domain of the socket for which to find a path.  This is the same as the
.I domain
argument to
.BR socket (3).
.TP
.I type
The type of the socket for which to find a path.  This is the same as the
.I type
argument to
.BR socket (3).
.TP
.I protocol
The protocol of the socket for which to find a path.  This is the same as the
.I protocol
argument to
.BR socket (3).
.TP
.I sockpath
A pointer to a
.B sockpath
structure returned from
.BR getsockpathent ()
to be freed by
.BR freesockpathent ().
.TP
.I msg
A short message string to prefix to the error message printed to standard
error by the
.BR sp_perror ()
subroutine.
.\"
.\"
.SH DESCRIPTION
.PP
.IR UNIX \(rg
.[
svr42
.]
has traditional used the
.BR netconfig (5s)
database for looking up
.I STREAMS
device paths for implementing sockets with the
.BR socklib (3)
library.
Later versions, particularly of
.IR Solaris \(rg
.[
soldoc
.]
have implemented a
.BR sock2path (5)
database that is used to populate kernel datastructures for use by the
.BR socksys (4)
driver.
Using
.BR netconfig (5s)
to look up socket paths is problematic in that it does not support the full
range of socket types, and in particular
.BR SOCK_SEQPACKET " and " SOCK_RDM .
This set of database lookup functions provides access to the
.BR sock2path (5)
database from user space in a similar manner to that of
.BR netconfig (3).
.PP
.BR setsockpath (),
.BR getsockpath (),
.BR endsockpath (),
.BR getsockpathent (),
.BR freesockpathent (),
.BR sp_perror "() and "
.BR sp_sperror (),
provide a set of Socket Path database access subroutines for accessing the
Socket Path database contained in the
.BR /etc/sock2path ,
.BR sock2path (5),
file.
.PP
.BR setsockpath (),
.BR getsockpath "() and "
.BR endsockpath (),
provide a mechanism to sequentially access each entry in the Socket Path
database.
.TP
.BR setsockpath ()
returns a pointer to an opaque datastructure used as a handle to the
.BR getsockpath ()
subroutine.
.BR setsockpath ()
is used to initialize the client access to the database entries and set the
access point to the first entry.
.TP
.BR getsockpath ()
when called successively with the
.I handle
returned by the call to
.BR setsockpath (),
returnes each Socket Path database entry starting with the first.  When no
additional entries remain that have not already been retrieved since the call
to
.BR setsockpath (),
.BR getsockpath ()
returns a
.B NULL
pointer to indicate that the end of the database has been reached.
.TP
.BR endsockpath ()
releases client access to the database and frees for reuse any resources that
were held by
.BR setsockpath ()
for use by
.BR getsockpath ().
.BR endsockpath ()
must only be called with a
.I handle
returned from a previous and corresponding call to
.BR setsockpath ().
.BR endsockpath ()
must be called for each call to
.BR setsockpath ()
with the corresponding
.I handle
pointer as the argument.
.PP
.BR getsockpathent "() and "
.BR freesockpathent (),
provide a mechanism to access the Socket Path database entries by
.BR socket (3)
.IR domain ,
.IR type " and "
.IR protocol .
.TP
.BR getsockpathent ()
Returns a pointer to the
.B sockpath
structure corresponding to the socket path entry with the same
.IR domain ", " type " and " protocol
as that supplied in the corresponding arguments.
.TP
.BR freesockpathent ()
frees the socket path entry returned as a pointer to a
.B sockpath
structure, from a previous call to
.BR getsockpathent ().
.BR freesockpathent ()
must be called for each call to
.BR getsockpathent ()
with the corresponding
.I sockpath
pointer as the argument.
.PP
.BR sp_perror "() and "
.BR sp_sperror ()
provide a mechanism for providing diagnostic messages upon failure.
.TP
.BR sp_perror ()
prints a diagnostic message to standard error with the string provided in the
.I msg
argument used as a prefix followed by a colon.  Diagnostic messages printed to
standard error are followed with a NEWLINE.
.TP
.BR sp_sperror ()
is similar to
.BR sp_perror (),
but, instead of printing the message to standard error, only a pointer to the
diagnostic message string is returned.  The allows the caller to subsequently
format and store a message according to the caller's specifications.
.\"
.\"

NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.


src/lib/sockpath.c
/*****************************************************************************

 @(#) File: src/lib/sockpath.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/lib/sockpath.c (" PACKAGE_ENVR ") " PACKAGE_DATE;

/* This file can be processed with doxygen(1). */

/*
 *  This is a simple library for accessing the /etc/sock2path file in a manner similar to the
 *  netconfig(3) functions.  It is used both by the soconfig(8) utility as well as the libsocklib
 *  library.
 */

#ifndef _XOPEN_SOURCE
#define _XOPEN_SOURCE 600
#endif
#define _REENTRANT
#define _THREAD_SAFE

#if 0
#define __USE_UNIX98
#define __USE_XOPEN2K
#define __USE_GNU
#endif

#include <stdlib.h>
#include "gettext.h"
#include <sys/types.h>
#include <sys/stat.h>
#ifdef HAVE_SYS_IOCTL_H
#include <sys/ioctl.h>
#endif
#include <sys/stropts.h>
#include <sys/poll.h>
#include <fcntl.h>

#ifdef HAVE_INTTYPES_H
# include <inttypes.h>
#else
# ifdef HAVE_STDINT_H
#  include <stdint.h>
# endif
#endif

#ifndef __EXCEPTIONS
#define __EXCEPTIONS 1
#endif

#include <unistd.h>
#include <errno.h>
#include <stddef.h>
#include <stdio.h>
#include <string.h>
#include <stropts.h>
#include <pthread.h>
#include <linux/limits.h>
#include <values.h>

#ifndef __P
#define __P(__prototype) __prototype
#endif

#include <sockdb.h>
#include "sockpath.h"

#if defined __i386__ || defined __x86_64__ || defined __k8__
#define fastcall __attribute__((__regparm__(3)))
#else
#define fastcall
#endif

#define __hot __attribute__((section(".text.hot")))
#define __unlikely __attribute__((section(".text.unlikely")))

#if __GNUC__ < 3
#define inline static inline fastcall __hot
#define noinline extern fastcall __unlikely
#else
#define inline static inline __attribute__((always_inline)) fastcall __hot
#define noinline static __attribute__((noinline)) fastcall __unlikely
#endif

#define likely(x) __builtin_expect(!!(x), 1)
#define unlikely(x) __builtin_expect(!!(x), 0)

#undef min
#define min(a, b) (a < b ? a : b)

/*
 *  Once condition for Thread-Specific Data key creation.
 */
static pthread_once_t __sockpath_tsd_once = PTHREAD_ONCE_INIT;

/*
 *  SOCKPATH library Thread-Specific Data key.
 */
static pthread_key_t __sockpath_tsd_key = 0;

static void
__sockpath_tsd_free(void *buf)
{
	pthread_setspecific(__sockpath_tsd_key, NULL);
	free(buf);
}

static void
__sockpath_tsd_key_create(void)
{
	pthread_key_create(&__sockpath_tsd_key, __sockpath_tsd_free);
}

static struct __sockpath_tsd *
__sockpath_tsd_alloc(void)
{
	struct __sockpath_tsd *tsdp;

	tsdp = (typeof(tsdp)) malloc(sizeof(*tsdp));
	memset(tsdp, 0, sizeof(*tsdp));
	pthread_setspecific(__sockpath_tsd_key, (void *) tsdp);
	return (tsdp);
}

static struct __sockpath_tsd *
__sockpath_get_tsd(void)
{
	struct __sockpath_tsd *tsdp;

	pthread_once(&__sockpath_tsd_once, __sockpath_tsd_key_create);
	if (unlikely((tsdp = (typeof(tsdp)) pthread_getspecific(__sockpath_tsd_key)) == NULL))
		tsdp = __sockpath_tsd_alloc();
	return (tsdp);
}

int *
__sockpath___sperror(void)
{
	return &(__sockpath_get_tsd()->sperror);
}

__asm__(".symver __sockpath___sperror,__sperror@@SOCKPATH_1.0");

/*
 *  Socket Path functions:
 *  ======================
 */

const char *__sockpath_sock2path = SOCK2PATH;

#undef SOCK2PATH
#define SOCK2PATH __sockpath_sock2path

static pthread_rwlock_t __sockpath_sp_lock = PTHREAD_RWLOCK_INITIALIZER;
static struct sockpathent **__sockpath_sp_list = NULL;

static struct sockpathent **
__sockpath_loadsockpathlist(void)
{
	char buffer[1024];
	char *line;
	int linenum, entries = 0;
	FILE *file = NULL;
	struct sockpathent *se = NULL, *selist = NULL, **selistp = &selist;
	struct __sockpath_tsd *tsd = __sockpath_get_tsd();

	if ((file = fopen(SOCK2PATH, "r")) == NULL)
		goto openfail;

	/* read file one line at a time */
	for (linenum = 1; (line = fgets(buffer, sizeof(buffer), file)) != NULL; linenum++) {
		char *str, *field = NULL, *tmp = NULL;
		int fieldnum;

		/* allocate one if we don't already have one */
		if (se == NULL) {
			if ((se = malloc(sizeof(*se))) == NULL)
				goto nomem;
			memset(se, 0, sizeof(*se));
		}
		for (str = line, fieldnum = 0; fieldnum < 4
		     && (field = strtok_r(str, " \t\f\n\r\v", &tmp)) != NULL;
		     str = NULL, fieldnum++) {
			char *end;

			switch (fieldnum) {
			case 0:
				se->se_sp.sp_domain = strtoul(field, &end, 0);
				if (end[0] != '\0' || field == end)
					break;
				continue;
			case 1:
				se->se_sp.sp_type = strtoul(field, &end, 0);
				if (end[0] != '\0' || field == end)
					break;
				continue;
			case 2:
				se->se_sp.sp_protocol = strtoul(field, &end, 0);
				if (end[0] != '\0' || field == end)
					break;
				continue;
			case 3:
				if ((se->se_sp.sp_path = strdup(field)) == NULL)
					break;
				continue;
			}
			break;
		}
		if (fieldnum == 0 && (field == NULL || field[0] == '#'))
			continue;	/* skip blank or comment lines */
		if (fieldnum != 4)
			goto error;
		/* good entry, add it to the list */
		*selistp = se;
		selistp = &se->se_next;
		se = NULL;
		entries++;
	}
	{
		struct sockpath **spp;
		int i;

		/* create master array */
		if ((spp = calloc(entries + 1, sizeof(struct sockpath *))) == NULL) {
			tsd->sperror = SP_NOMEM;
			goto error;
		}
		for (se = selist, i = 0; i < entries; i++) {
			spp[i] = (struct sockpath *) se;
			se = se->se_next;
		}
		spp[entries] = NULL;

		return ((__sockpath_sp_list = (struct sockpathent **) spp));
	}
      nomem:
	tsd->sperror = SP_NOMEM;
	goto error;
      openfail:
	tsd->sperror = SP_OPENFAIL;
	goto error;
      error:
	if (se != NULL) {
		*selistp = se;
		selistp = &se->se_next;
		se = NULL;
	}
	while ((se = selist)) {
		selist = se->se_next;
		if (se->se_sp.sp_path)
			free(se->se_sp.sp_path);
		free(se);
	}
	if (file)
		fclose(file);
	return (NULL);
}

static struct sockpathent **
__sockpath_getsockpathlist(void)
{
	struct sockpathent **sep;

	pthread_rwlock_rdlock(&__sockpath_sp_lock);
	if ((sep = __sockpath_sp_list) != NULL) {
		pthread_rwlock_unlock(&__sockpath_sp_lock);
		return (sep);
	}
	pthread_rwlock_unlock(&__sockpath_sp_lock);
	pthread_rwlock_wrlock(&__sockpath_sp_lock);
	if ((sep = __sockpath_sp_list) != NULL) {
		pthread_rwlock_unlock(&__sockpath_sp_lock);
		return (sep);
	}
	sep = __sockpath_loadsockpathlist();
	pthread_rwlock_unlock(&__sockpath_sp_lock);
	return (sep);
}

/**
 * @fn void *setsockpath(void)
 * @brief Provide handle to the sock2path database.
 *
 * This function has the effect of binding to or rewinding the sock2path databse.  This function
 * must be called before the first call to getsockpath() and may be called at any other time.  This
 * function need no be called before a call to getsockpathent().  This function returns NULL at the
 * end of the file, or upon failure.
 */
void *
__sockpath_setsockpath(void)
{
	struct sockpathent **sep;
	struct sockpath_handle *sh = NULL;

	if ((sep = __sockpath_getsockpathlist()) == NULL)
		return (NULL);
	if ((sh = malloc(sizeof(*sh))) == NULL) {
		sp_error = SP_NOMEM;
		return (NULL);
	}
	pthread_rwlock_init(&sh->sh_lock, NULL);
	sh->sh_head = sep;
	sh->sh_curr = sep;
	sp_error = SP_NOERROR;
	return ((void *) sh);
}

__asm__(".symver __sockpath_setsockpath,setsockpath@@SOCKPATH_1.0");

/**
 * @fn struct sockpath *getsockpath(void *handle)
 * @brief Retrieve next entry in the sock2path database.
 * @param handle handle returned by setsockpath().
 *
 * This function returns a pointer to the current entry in the sock2path database, formatted as a
 * struct sockpath.  Successive calls will return successive sock2path entries in the sock2path
 * databse.  This function can be used to search the entire sock2path file.  This function returns
 * NULL at the end of the file.  handle is the handle obtained through setsockpath().
 */
struct sockpath *
__sockpath_getsockpath(void *handle)
{
	struct sockpath_handle *sh = (struct sockpath_handle *) handle;
	struct sockpath *sp;

	pthread_rwlock_rdlock(&__sockpath_sp_lock);
	if (__sockpath_sp_list == NULL || sh == NULL) {
		sp_error = SP_NOSET;
		pthread_rwlock_unlock(&__sockpath_sp_lock);
		return (NULL);
	}
	pthread_rwlock_unlock(&__sockpath_sp_lock);
	pthread_rwlock_wrlock(&sh->sh_lock);
	if (!(sp = (struct sockpath *) (*(sh->sh_curr)))) {
		sp_error = SP_NOMOREENTRIES;
		pthread_rwlock_unlock(&sh->sh_lock);
		return (NULL);
	}
	sh->sh_curr++;
	sp_error = SP_NOERROR;
	pthread_rwlock_unlock(&sh->sh_lock);
	return (sp);
}

__asm__(".symver __sockpath_getsockpath,getsockpath@@SOCKPATH_1.0");

/**
 * @fn int endsockpath(void *handle)
 * @brief Release socket path database.
 * @param handle handle returned by setsockpath().
 *
 * This function should be called when processing is complete to release resources held for reuse.
 * handle is the handle obtained through setsockpath().  Programmers should be aware, however, that
 * the last call to endsockpath() frees all memory allocated by getsockpath() for the struct
 * sockpath data structures.  This function may not be called before setsockpath().
 */
int
__sockpath_endsockpath(void *handle)
{
	struct sockpath_handle *sh = (struct sockpath_handle *) handle;

	pthread_rwlock_rdlock(&__sockpath_sp_lock);
	if (__sockpath_sp_list == NULL || sh == NULL) {
		sp_error = SP_NOSET;
		pthread_rwlock_unlock(&__sockpath_sp_lock);
		return (-1);
	}
	pthread_rwlock_unlock(&__sockpath_sp_lock);
	/* strange thing to do, no? */
	pthread_rwlock_wrlock(&sh->sh_lock);
	free(sh);
	sp_error = SP_NOERROR;
	return (0);
}

__asm__(".symver __sockpath_endsockpath,endsockpath@@SOCKPATH_1.0");

/**
 * @fn struct sockpath *getsockpathent(int domain, int type, int protocol)
 * @brief Return a socket path database entry for a domain, type and protocol.
 * @param domain domain from the socket(3) call.
 * @param type type from the socket(3) call.
 * @param protocol protocol from the socket(3) call.
 *
 * This function returns a pointer to the struct sockpath structure corresponding to the
 * arguments domain, type and protocol.  It returns NULL if any argument is invalid (that is, does
 * not have an entry in the socket path database).
 *
 * This function returns a pointer to the struct sockpath structure on success and NULL on
 * failure.
 */
struct sockpath *
__sockpath_getsockpathent(int domain, int type, int protocol)
{
	struct sockpathent **sep;
	struct sockpath *sp = NULL;

	if ((sep = __sockpath_getsockpathlist())) {
		sp_error = SP_NOTFOUND;
		for (; *sep; ++sep) {
			if (((*sep)->se_sp.sp_domain == domain || domain == 0) &&
			    ((*sep)->se_sp.sp_type == type) &&
			    ((*sep)->se_sp.sp_protocol == protocol || protocol == 0)) {
				if ((sp = malloc(sizeof(*sp))) == NULL) {
					sp_error = SP_NOMEM;
					break;
				}
				memcpy(sp, (*sep), sizeof(*sp));
				if ((sp->sp_path = strdup((*sep)->se_sp.sp_path)) == NULL) {
					free(sp);
					sp_error = SP_NOMEM;
					break;
				}
			}
			sp_error = SP_NOERROR;
			break;
		}
	}
	return (sp);
}

__asm__(".symver __sockpath_getsockpathent,getsockpathent@@SOCKPATH_1.0");

/**
 * @fn void freesockpathent(struct sockpath *sp)
 * @brief Free a socket path database entry.
 * @param sp the database entry to free.
 *
 * This function frees the sockpath structure pointed to by sp (previously returned by
 * getsockpathent()).
 */
void
__sockpath_freesockpathent(struct sockpath *sp)
{
	if (sp) {
		if (sp->sp_path)
			free(sp->sp_path);
		free(sp);
	}
}

__asm__(".symver __sockpath_freesockpathent,freesockpathent@@SOCKPATH_1.0");

/* *INDENT-OFF* */
static const char *__sockpath_sp_errlist[] = {
/*
TRANS Corresponds to the SP_NOERROR constant.  No error is indicated in the
TRANS nc_error(3) variable.  The last sockpath(3) operation was successful.
 */
	[SP_NOERROR] = gettext_noop("no error"),
/*
TRANS Corresponds to the SP_NOMEM constant.  Out of memory.  Memory could not be
TRANS allocated to complete the last sockpath(3) operation.
 */
	[SP_NOMEM] = gettext_noop("out of memory"),
/*
TRANS Corresponds to the SP_NOSET constant.  The getsockpath(3) or
TRANS endsockpath(3) function was called out of order: that is, before
TRANS setsockpath(3) was called.
 */
	[SP_NOSET] = gettext_noop("routine called out of order"),
/*
TRANS Corresponds to the SP_OPENFAIL constant.  The /etc/sock2path file could
TRANS not be opened for reading.  %1$s is the name of the /etc/sock2path file.
 */
	[SP_OPENFAIL] = gettext_noop("could not open %s"),
/*
TRANS Corresponds to the SP_BADLINE constant.  A syntax error exists in the
TRANS /etc/sock2path file at the location shown.  The field number is the
TRANS whitespace separated field within the line in the /etc/sock2path file at
TRANS which the syntax error was detected.  The line number is the line number
TRANS in the /etc/sock2path file at which the syntax error was detected.  %1$s
TRANS is the name of the /etc/sock2path file.  %2$d is the field number.  %3$d
TRANS is the line number.
 */
	[SP_BADLINE] = gettext_noop("syntax error in %s: field %d of line %d"),
/*
TRANS Corresponds to the SP_NOTFOUND constant.  The "domain", "type" or
TRANS "protocol" specified as an argument to the getsockpathent(3) subroutine
TRANS was not found in any entry in the /etc/sock2path file.  %1$s is the name
TRANS of the /etc/sock2path file.
 */
	[SP_NOTFOUND] = gettext_noop("netid not found in %s"),
/*
TRANS Corresponds to the SP_NOMOREENTRIES constant.  All entries in the
TRANS /etc/sock2path file has been exhausted.  They have either already been
TRANS retrieved with getsockpath(3), getsockpathent(3), or the /etc/sock2path
TRANS file contains no entries.  %1$s is the name of the /etc/sock2path file.
 */
	[SP_NOMOREENTRIES] = gettext_noop("no more entries in %s"),
/*
TRANS Corresponds to any other constant.  An internal error in the libxnsl(3)
TRANS library has occurred resulting in an undefined error number.  This
TRANS situation should not occur and represents a bug in the library.  %1$d is
TRANS the decimal value of the unknown error number.
 */
	[SP_ERROR_MAX] = gettext_noop("unknown error number %d"),
};
/* *INDENT-ON* */

/**
 * @fn char *sp_sperror(void)
 * @brief Return and error string.
 *
 * This function is similar to sp_perror() but instead of printing the message to standard error,
 * will return a pointer to a string that contains the error message.
 *
 * This function returns a pointer to a buffer that contains the error message string.  This buffer
 * is overwritten on each call.  In multithreaded applications, this bufer is implemented as
 * thread-specific data.
 */
char *
__sockpath_sp_sperror(void)
{
	int err, idx;
	struct __sockpath_tsd *tsd = __sockpath_get_tsd();
	char *errbuf = tsd->sperrbuf;

	if (errbuf != NULL) {
		switch ((idx = err = tsd->sperror)) {
		case SP_NOERROR:
		case SP_NOMEM:
		case SP_NOSET:
			(void) strncpy(errbuf, __sockpath_sp_errlist[idx], SPERR_BUFSZ);
			break;
		case SP_OPENFAIL:
		case SP_NOTFOUND:
		case SP_NOMOREENTRIES:
			(void) snprintf(errbuf, SPERR_BUFSZ, __sockpath_sp_errlist[idx], SOCK2PATH);
			break;
		case SP_BADLINE:
			(void) snprintf(errbuf, SPERR_BUFSZ, __sockpath_sp_errlist[idx], SOCK2PATH,
					tsd->fieldnum, tsd->linenum);
			break;
		default:
			idx = SP_ERROR_MAX;
			(void) snprintf(errbuf, SPERR_BUFSZ, __sockpath_sp_errlist[idx], err);
			break;
		}
	}
	return (errbuf);

}

__asm__(".symver __sockpath_sp_sperror,sp_sperror@@SOCKPATH_1.0");

/**
 * @fn void sp_perror(const char *msg)
 * @brief Print an error message to standard error.
 * @param msg message to prefix to error message.
 *
 * This function prints an error message to standard error indicating why any of the above routines
 * failed.  The message is prepended with the string provided in the msg argument plus a colon.  A
 * NEWLINE is appended to the end of the message.
 */
void
__sockpath_sp_perror(const char *msg)
{
	if (msg)
		fprintf(stderr, "%s: %s\n", msg, __sockpath_sp_sperror());
	else
		fprintf(stderr, "%s\n", __sockpath_sp_sperror());
	return;
}

__asm__(".symver __sockpath_sp_perror,sp_perror@@SOCKPATH_1.0");

/*
 * vim: comments+=b\:TRANS
 */
